### ISSUE TITLE
Modify SVG icon handling so that they can be used in CSS

### ISSUE THREAD

**Issue by Scotchester** [wagtail/wagtail#7511]
### Is your proposal related to a problem?

Some of the last icons that need to be migrated from our icon font to SVG (see: #6107) are currently inserted via CSS `content` properties.

In taking a look at the way these are used, it doesn't seem like these can be addressed by our preferred method of inlining the SVGs in the template, so we need to figure out a way to keep inserting them via CSS.

We had thought maybe it’d be simplest just to create data URLs in each case and plop them in there, but when I realized how many occurrences there are (and how many of them use the same icons), I thought we really should try to come up with a solution that loads the icons by reference so an update to the source SVG updates it in the CSS uses, as well.

Following discussion with @thibaudcolas, @allcaps, and the Accessibility Team, we devised a plan for how to do that.

### Describe the solution you'd like

1. Modify the source SVG files to use `svg` root elements, rather than `symbol`
2. Convert these SVGs to symbols for the purpose of creating a sprite in the same format as is currently being used ~~(presumably as a new part of the build process)~~
3. Serve the original SVGs as static files so they can be referenced from CSS

### Describe alternatives you've considered

One alternative option I threw out was to figure out some way to drop the contents of one of the existing icon SVGs (just the `path`s, no `symbol`) into the CSS to create a data URL. It seems like this could be done either before or after the Sass compilation. E.g., something like `content: %% SOME MAGIC ICON INVOCATION %%;` results in `content: url("data:image/svg+xml, <svg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 16 16' fill='gray'><path d='M12.12 14.592h-8.24c-… 2 0.412z'></path></svg>");` in the compiled CSS.

We agreed that this was too complicated, since it has to work around the Sass compilation and results in much messier Sass source files and compiled CSS.

### Additional context

- One downside to this approach is that SVG files brought in to CSS via `url()` (either external or as data URLs) cannot be styled. It appears that the best workaround for this is to [use the CSS `mask` property](https://css-tricks.com/change-color-of-svg-on-hover/#use-a-mask-instead-of-a-background-image) in combination with [`background-color: currentColor`](https://stackoverflow.com/a/66093928).
- We'll want to ensure the SVGs are properly cached so that there aren't a bunch of full requests made every time one of these CSS-referenced SVGs is needed.

**Comment by thibaudcolas** [comment]
FYI @Scotchester @stevedya @allcaps I believe we’ll proceed with this next week.

**Comment by thibaudcolas** [comment]
Note for future reference @Scotchester has implemented the first two points from this issue in #8222. The only thing left is:

> Serve the original SVGs as static files so they can be referenced from CSS

For this to work, we may have to add an extra `xmlns="http://www.w3.org/2000/svg"` to those `<svg>` elements (TBC).

**Comment by Lovelyfin00** [comment]
Hi @thibaudcolas
Has this been fixed?

**Comment by thibaudcolas** [comment]
@Lovelyfin00 there is still work needed – the third point from the original solution described by Scott. We need to:

- [ ] Serve the existing icons folder as static files (https://github.com/wagtail/wagtail/tree/main/wagtail/admin/templates/wagtailadmin/icons)
- [ ] Update the path at https://github.com/wagtail/wagtail/blob/main/client/scss/settings/_variables.scss#L9 so we now load the icons from that folder (depending on how it’s served as static files)
- [ ] Remove the icons we’ve duplicated here: https://github.com/wagtail/wagtail/tree/main/wagtail/admin/static_src/wagtailadmin/images/icons

And that should be it. I don’t have a good sense of what would be the best way to do the first step though – whether we need to add the folder to our static files’ collections, whether that means we need to move it out of templates or not, or whether we create a one-off static-files-serving route. Since Wagtail will live inside other people’s Django projects, it’s important we get this right. @Scotchester @zerolab any thoughts on how to do this correctly?

**Comment by gasman** [comment]
@thibaudcolas I don't think we can safely circumvent the requirement for static files to live in the `static` subfolder of an app. Storing these icons as templates has always felt like a bit of a hack to me - if I remember right, all we're really doing within the templating engine is repurposing `{% include %}` as a way of concatenating text files together, and we'd probably be better off doing that concatenation explicitly in Python code. If we were to do that, we could move them from `templates` to `static`.

We'll need to consider what happens with the `register_icons` hook though, since anything returned from that hook is currently expected to be a template path. `register_icons` is undocumented, so it _should_ be fair game to make breaking changes to it, but we should be conscious of the possibility of breaking wagtailfontawesome or similar. And if we do want to support third-party apps using that hook in future, we'll need to tap into Django's static file finder mechanism to locate those icon files when building the concatenated file (i.e. we can't just treat the paths we get from `register_icons` as local paths within the wagtailadmin app).

**Comment by zerolab** [comment]
fwiw, we use `register_icons` in several places outside of the admin (such as https://github.com/allcaps/wagtail-font-awesome-svg, wagtail-localize, and a number of projects that relied on wagtailfontawesome) and it is "documented" in #6028 and it was something we've pointed at. On that basis, it would be great to keep breaking changes at a minimum if possible.

Will have a think tomorrow/over the weekend, especially from the point of view as a package maintainer

**Comment by thibaudcolas** [comment]
We moved that concatenation to happen in Python already, I believe as part of Scott’s #8222: https://github.com/wagtail/wagtail/blob/0b4302afe9c1d4e23f84bf886ac5ff5c592d3928/wagtail/admin/views/home.py#L331-L341

Really all we’re using the template system for at the moment is the `render_to_string` convenience method to load the files’ contents.

I’ve not worked with Django’s static files APIs aside from serving the files, and only have vague awareness of the difference between the different static file storage backends. So would appreciate any further guidance on this!

---

Note using SVG files for CSS like this is very much a workaround we’d like to use as little as possible (it should only be needed for markup which we don’t control like jQuery Datetimepicker, or Django’s built-in widgets). @Scotchester I wonder if we should consider a completely different approach, for example serve the existing "inline SVG" sprite as an external sprite where we could use [SVG fragment identifiers](https://caniuse.com/svg-fragment) to target the specific icons?

**Comment by Scotchester** [comment]
> @Scotchester I wonder if we should consider a completely different approach, for example serve the existing "inline SVG" sprite as an external sprite where we could use [SVG fragment identifiers](https://caniuse.com/svg-fragment) to target the specific icons?

Wow! Another fancy SVG technique that has apparently worked forever that I was totally unaware of 😆 That looks very promising!

**Comment by Lovelyfin00** [comment]
@thibaudcolas
I guess this issue still needs a lot of decisions to be made on how to solve it☺☺

**Comment by thibaudcolas** [comment]
@Lovelyfin00 at this stage I think what would help is further exploration of the technique I mention in https://github.com/wagtail/wagtail/issues/7511#issuecomment-1286684140, as I think it’d solve a lot of problems for comparatively little effort – but it’s a technique so rarely used that I have no idea how ready it is for real-world usage. Once someone has done the experimentation, we can hopefully make an informed decision.

---

Here is a quick prototype I put together: https://wagtail-7511-svg-views.netlify.app/. Most of the magic is in that site’s [`/views.svg`](https://wagtail-7511-svg-views.netlify.app/) file.

This file is based on our existing icon "sprite", this time definitely a sprite in the traditional sense of the name: where a single image is cut into separate areas based on the position of different assets. It needs one `symbol`, one `use`, and one [`<view>`](https://developer.mozilla.org/en-US/docs/Web/SVG/Element/view) element per icon:

```svg
<svg xmlns="http://www.w3.org/2000/svg">
<defs>
<symbol id="symbol-warning"  viewBox="0 0 512 512"><path d="M506.3 417 293 53c-16.33-28-57.54-28-73.98 0L5.82 417c-16.41 27.9 4.029 63 36.92 63h426.6c32.76 0 53.26-35 36.96-63zM232 168c0-13.25 10.75-24 24-24s24 10.8 24 24v128c0 13.25-10.75 24-23.1 24S232 309.3 232 296V168zm24 248c-17.36 0-31.44-14.08-31.44-31.44s14.07-31.44 31.44-31.44 31.44 14.08 31.44 31.44C287.4 401.9 273.4 416 256 416z"/></symbol>
</defs>
<use href="#symbol-warning"  x="127000" y="127000" />
<view id="icon-warning"  viewBox="127000 127000 512 512"/>
</svg>
```

Here, for simplicity’s sake we define a `symbol` for each icon like our existing sprite. Then we place it somewhere in the image with `use` – so far so good. Finally we use [`<view>`](https://developer.mozilla.org/en-US/docs/Web/SVG/Element/view) to create a [SVG fragment identifier](https://caniuse.com/svg-fragment) which we can reference from outside the file.

The three experimental aspects that I think need further validation are:

- The `<view>` element (browser support, any other concerns)
- SVG fragment identifiers (browser support in `background-image` and `mask-image`)
- Our ability to automatically place icons within a sprite of `<view>` elements. We need to assign each icon a pair of x-y coordinates where the whole icon’s `viewBox` can be displayed without overlap with any other icon. Here I assigned each icon a 1000x1000 area along the diagonal of the SVG horizontal and vertical axis only, for simplicity. That feels pretty wasteful. though I’m not sure there’s any actual perf issues with large SVG coordinates (with 127 icons, the whole SVG’s viewBox is `0 0 127512 127512`)

**Comment by allcaps** [comment]
On the templates vs static files: I fail to see what benefit static files give us. We do not need to serve individual icons. We process the individual icons and turn them into a single sprite. The sprite is a single Django view. To process static files in this no-default way is confusing and a hack too.

The way staticfiles app does file look-ups is very similar to template look-ups, but static files also gives additional complexity that we don't need. Like collecting static and serving the individual icons.

I'm the author of [Wagtail Font Awesome SVG](https://github.com/wagtail-nest/wagtail-font-awesome-svg). It is created as a drop-in replacement for [Wagtail Font Awesome](https://pypi.org/project/wagtailfontawesome/)  (the icon font). If Wagtail decides to change the implementation, I'm happy to update Font Awesome SVG.

**Comment by thibaudcolas** [comment]
@allcaps serving individual icons is very desirable in specific cases. A good example is the custom arrow for `<select>` elements, where we’re relying on the markup from Django rather than having our custom SelectWithArrow widget.

**Comment by allcaps** [comment]
Okay, for serving icons individually, we don't need a view in Wagtail. Then icons are best stored and served as regular static files. No doubt. But, can we do without SVG to symbol conversion? Can we do without a sprite?

Just thinking out loud here:

We could create a management command `createicons` to convert icons to various output formats. Wagtail could read from, and write to static directories.

This management command could be part of the release process.  So we can ship the defaults with the Wagtail package.

If a integrator or third-party adds a custom icon, `createicons` needs to be executed before `collecstatic` to update the static files.

The main benefit would be that we can create multiple output formats to support various techniques. Eg: individual SVGs files, CSS file with CSS variables, SVG symbol sprite, JS components, etc.

**Comment by thibaudcolas** [comment]
Picking this back up – after having explored a bit more what’s possible, there are four different approaches that seem worth considering further. I could use some help to figure out which one to go for:

1. Keep doing what we do now – duplicate icon source files only for icons that need to be used in CSS.
2. A `npm run createicons` bash or Node script to generate/copy the SVG files.
3. A "`createicons`" Django management command to generate/copy the SVG static files (as described by Coen above)
4. A Django view that serves the icons as needed for CSS references

## 1. Duplicate icon source files

This works ok. It means needing to be aware _some_ icon files have to be duplicated, and where to add them, but nothing we can’t document, and works reasonably well for a small number of icons (referencing icons from CSS should only be a last resort anyway).

- It’s very simple to set up, understand, and troubleshoot: still just leveraging normal static files management in Django
- There is no magic involved and it’s all manual so needs to be something people remember to do when they update icons that need referencing in CSS.
- Per-project icon overrides are straightforward – people just have to place their overriding file in the right location with the right file name (just like template overrides)
- This would be completely separately from our `register_icons` hook, so really only valid as a workaround for where icons _must_ be set in CSS.

## 2. `createicons` script

Slight improvement on the above – rather than duplicating the files in source control, we just have a build script that copies all the files over to the right folder. Running the equivalent of the following command as part of our `npm run build`:

```bash
cp -R wagtail/admin/templates/wagtailadmin/icons wagtail/admin/static/wagtailadmin/images/icons
```

## 3. `createicons` management command

This is a more complex iteration on the above, that would allow us to have _some_ magic, such as awareness of `register_icons`. Not sure whether the complexity is worth it considering CSS references to icons are a last resort, but it’s worth considering it. This still has the disadvantage of being a manual step people will have to remember to do when updating icons.

## 4. Django view

This is similar to what we’ve done currently for icons usage in HTML. This view would have to create a similar "sprite", with SVG `<view>` elements in addition to the symbols (see [prototype](https://wagtail-7511-svg-views.netlify.app/)). This can be fully aware of the `register_icons` hook, and won’t require any further manual step when adding or removing icons.

This was my preferred approach until a few hours ago – since this is a regular view, we can’t just hard-code its URL in the stylesheets like we can for SVG static files. So all icons used this way would need to be referenced via a CSS variable set via JavaScript, so we know the correct view URL. Something like:

```js
document.documentElement.style.setProperty('--w-icon-angle-double-left', 'http://localhost:8000/admin/icons-views-sprite-365bdee8/')

// And then in CSS:
// background-image: url(var(--w-icon-angle-double-left));
```

---

Unless told otherwise I think I’ll go further with the existing "Duplicate icon source files" approach. We use this approach for 7 icons, and if my analysis is correct there should only be about 5 more icons needing this, all of which are for the jQuery Datetimepicker which we want to eventually remove anyway (#8056).

**Comment by thibaudcolas** [comment]
We’ve gone very far with the "duplicate icons" approach, and in the end this is more or less only needed for legacy code. So I’ll close this now.
